service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryMapper.java

```diff
@@ -0,0 +1,161 @@
+            var agentDataList = AgentDirectoryExtractor.extractAgentData(bundle, List.of((Practitioner) practitioner.get()))
+                .stream()
+                .map(agentData -> mapAgentPerson(agentData.getPractitioner(), agentData.getPractitionerRole(), agentData.getOrganization()))
+                .filter(StringUtils::isNotEmpty)
+                .findFirst();
+            if (agentDataList.isPresent()) {
+                return agentDataList.get();
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String buildAgentPerson(ReferralRequest referralRequest, Bundle bundle) {
+        var practitioner = ResourceExtractor
+            .extractResourceByReference(bundle, referralRequest.getRequester().getAgent().getReferenceElement());
+        var organization = ResourceExtractor
+            .extractResourceByReference(bundle, referralRequest.getRequester().getOnBehalfOf().getReferenceElement());
+        if (practitioner.isPresent() && organization.isPresent()) {
+            return mapAgentPerson(practitioner.map(Practitioner.class::cast).get(),
+                null,
+                organization.map(Organization.class::cast).get());
+
+        } else if (practitioner.isPresent()) {
+            var agentDataList = AgentDirectoryExtractor.extractAgentData(bundle, List.of(practitioner.map(Practitioner.class::cast).get()))
+                .stream()
+                .map(agentData -> mapAgentPerson(agentData.getPractitioner(), agentData.getPractitionerRole(), agentData.getOrganization()))
+                .filter(StringUtils::isNotEmpty)
+                .findFirst();
+
+            if (agentDataList.isPresent()) {
+                return agentDataList.get();
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String mapAgentPerson(Practitioner practitioner, PractitionerRole practitionerRole, Organization organization) {
+        if (!organizationPractitionerHasBeenMapped(practitioner, organization)) {
+            addOrganizationPractitionerPairToMap(practitioner, organization);
+            return practitionerAgentPersonMapper.mapPractitionerToAgentPerson(
+                practitioner,
+                Optional.ofNullable(practitionerRole),
+                Optional.ofNullable(organization)
+            );
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private boolean organizationPractitionerHasBeenMapped(Practitioner practitioner, Organization organization) {
+        var mappedId = createMappedId(practitioner, organization);
+        return mappedOrganizationsAndPractitioner.contains(mappedId);
+    }
+
+    private void addOrganizationPractitionerPairToMap(Practitioner practitioner, Organization organization) {
+        var mappedId = createMappedId(practitioner, organization);
+        mappedOrganizationsAndPractitioner.add(mappedId);
+    }
+
+    private String createMappedId(Practitioner practitioner, Organization organization) {
```

Comment:
This is still being called twice (if organisation and Practitioner have not been mapped)